File: Java-Core/src/main/java/presto/expression/SliceSelector.java
package presto.expression;

import presto.error.NullReferenceError;
import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.runtime.Context;
import presto.type.IType;
import presto.type.IntegerType;
import presto.utils.CodeWriter;
import presto.value.ISliceable;
import presto.value.IValue;
import presto.value.Integer;

public class SliceSelector extends SelectorExpression {

	IExpression first;
	IExpression last;
	
	public SliceSelector(IExpression first, IExpression last) {
		this.first = first;
		this.last = last;
	}
	
	public SliceSelector(IExpression parent, IExpression first, IExpression last) {
		super(parent);
		this.first = first;
		this.last = last;
	}

	public IExpression getFirst() {
		return first;
	}
	
	public IExpression getLast() {
		return last;
	}

	@Override
	public void toDialect(CodeWriter writer) {
		parent.toDialect(writer);
		writer.append('[');
		if(first!=null)
			first.toDialect(writer);
		writer.append(':');
		if(last!=null)
			last.toDialect(writer);
		writer.append(']');
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		IType firstType = first!=null ? first.check(context) : null;
		IType lastType = last!=null ? last.check(context) : null;
		if(firstType!=null && !(firstType instanceof IntegerType))
			throw new SyntaxError(firstType.toString() + " is not an integer");
		if(lastType!=null && !(lastType instanceof IntegerType))
			throw new SyntaxError(lastType.toString() + " is not an integer");
		IType parentType = parent.check(context);
		return parentType.checkSlice(context);
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
        IValue o = parent.interpret(context);
        if (o == null)
            throw new NullReferenceError();
        ISliceable<IValue> sliceable = o.asSliceable(context);
        if (sliceable!=null)
        {
            Object fi = first != null ? first.interpret(context) : null;
            if (fi != null && !(fi instanceof Integer))
                throw new SyntaxError("Illegal sliced type: " + fi);
            Object li = last != null ? last.interpret(context) : null;
            if (li != null && !(li instanceof Integer))
                throw new SyntaxError("Illegal sliced type: " + li);
            return sliceable.slice((Integer)fi, (Integer)li);
        }
        else
			throw new SyntaxError("Illegal sliced object: " + parent);
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/expression/SliceSelector.js
var SelectorExpression = require("./SelectorExpression").SelectorExpression;
var NullReferenceError = require("../error/NullReferenceError").NullReferenceError;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var IntegerType = require("../type/IntegerType").IntegerType;
var Integer = require("../value/Integer").Integer;

function SliceSelector(parent, first, last) {
	SelectorExpression.call(this, parent);
	this.first = first || null;
	this.last = last || null;
	return this;
}

SliceSelector.prototype = Object.create(SelectorExpression.prototype);
SliceSelector.prototype.constructor = SliceSelector;

SliceSelector.prototype.toString = function() {
	return this.parent.toString() + "[" +
			(this.first==null?"":this.first.toString()) + ":" +
			(this.last==null?"":this.last.toString()) + "]";
};
SliceSelector.prototype.toDialect = function(writer) {
    this.parent.toDialect(writer);
    writer.append('[');
    if (this.first != null)
        this.first.toDialect(writer);
    writer.append(':');
    if (this.last != null)
        this.last.toDialect(writer);
    writer.append(']');
};

SliceSelector.prototype.check = function(context) {
	var firstType = this.first!=null ? this.first.check(context) : null;
	var lastType = this.last!=null ? this.last.check(context) : null;
	if(this.firstType!=null && !(this.firstType instanceof IntegerType)) {
		throw new SyntaxError(this.firstType.toString() + " is not an integer");
	}
	if(this.lastType!=null && !(this.lastType instanceof IntegerType)) {
		throw new SyntaxError(this.lastType.toString() + " is not an integer");
	}
	var parentType = this.parent.check(context);
	return parentType.checkSlice(context);
};

SliceSelector.prototype.interpret = function(context) {
	var o = this.parent.interpret(context);
	if (o == null) {
		throw new NullReferenceError();
	}
    if (o.sliceable)
        o = o.sliceable;
	if (o.slice) {
		var fi = this.first != null ? this.first.interpret(context) : null;
		if (fi != null && !(fi instanceof Integer)) {
			throw new SyntaxError("Illegal slice value type: " + fi);
		}
		var li = this.last != null ? this.last.interpret(context) : null;
		if (li != null && !(li instanceof Integer)) {
			throw new SyntaxError("Illegal slice value type: " + li);
		}
		return o.slice(fi, li);
	} else {
		throw new SyntaxError("Illegal sliced object: " + this.parent);
	}
};

exports.SliceSelector = SliceSelector;
